From f0ec5c4892899f8236c71b9f3a461cc5d58647fe Mon Sep 17 00:00:00 2001
From: Brandon Maier <brandon.maier@gmail.com>
Date: Thu, 19 Dec 2024 20:36:42 -0600
Subject: [PATCH aerc] config: fix system paths when relocating aerc

The following error occurs when packaging aerc for Conda:

  $ aerc
  /home/conda/.config/aerc/aerc.conf not found, installing the system default
  error: open /usr/share/aerc/aerc.conf: no such file or directory

Conda installs aerc to a custom $PREFIX, so the config files will not be
available in /usr/share/aerc. The above error occurs because aerc
searches multiple paths for config paths and /usr/share/aerc happens to
be the last path. The $PREFIX is first in the search list, so we would
expect it to find the Conda installed files.

By dumping out the contents of the search paths, I observed the $PREFIX
paths are being searched but they are corrupted because their strings
end with embedded NULLs (\x00).

Conda builds aerc with a $PREFIX that is extremely long. Then when
installing aerc, it does relocation by replacing the build $PREFIX with
the final install $PREFIX, appending NULLs to fill out the string.

go-lang does not use NULL-terminated strings, so these NULLs are
preserved in the path. Trim any NULLs from the end of $PREFIX paths.

Signed-off-by: Brandon Maier <brandon.maier@gmail.com>

---
v1: https://lists.sr.ht/~rjarry/aerc-devel/patches/56556
---
 config/config.go | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/config/config.go b/config/config.go
index 6bab5fc8..f5230e31 100644
--- a/config/config.go
+++ b/config/config.go
@@ -34,9 +34,12 @@ func buildDefaultDirs() []string {
 	}
 
 	// Add custom buildtime dirs
+	libexecDir := strings.TrimRight(libexecDir, "\x00")
 	if libexecDir != "" && libexecDir != "/usr/local/libexec/aerc" {
 		defaultDirs = append(defaultDirs, xdg.ExpandHome(libexecDir))
 	}
+
+	shareDir := strings.TrimRight(shareDir, "\x00")
 	if shareDir != "" && shareDir != "/usr/local/share/aerc" {
 		defaultDirs = append(defaultDirs, xdg.ExpandHome(shareDir))
 	}
-- 
2.47.1

